### PR TITLE
XWIKI-20907: Introduce the notion of required rights

### PR DESCRIPTION
**Pull request by michitux**
# Jira URL

https://jira.xwiki.org/browse/XWIKI-20907

# Changes

## Description

* Add a new flag to XWikiDocument if required rights shall be enforced.
* Add the new flag to the filter stream and XAR APIs, increase the XAR version and adapt tests.
* Add the new flag to the REST API.
* Add the new flag to the edit form to support updating it.
* Add a DocumentRequiredRightsManager API to allow getting the required rights that are set on a document.
* Add a DocumentAuthorizationManager to check rights using required rights.
* Restrict edit right to users that have all required rights.

TODO:
* [x] Add tests
* [x] Fix the resolving of implied rights
* [ ] Add at least some UI, even if initially only for advanced users
* [x] Integrate required right checking into the contextual authorization manager for script and programming right.
* [x] Change as many existing right checks as possible to take document rights into account (e.g., UIX, translations, …)
* [x] Try to find a way to take required rights into account when saving documents with script rights
* [x] Address the remaining TODOs in the new code
* [ ] Add a cache for required rights (in `DocumentRequiredRightsManager`)
* [x] Change the versions to 16.10.0RC1 as this is not ready for 16.6.0.
* [x] Add a mandatory class initializer for the `XWiki.RequiredRightClass`

## Clarifications

* See https://design.xwiki.org/xwiki/bin/view/Proposal/Documentrightsmanagementimprovement for the design decisions taken.

# Screenshots & Video

# Executed Tests

Ran tests on all modules with code changes without quality profile (coverage is not met currently, this will be fixed).

# Expected merging strategy

* Prefers squash: No, these are huge changes. I tried to group them into somewhat meaningful commits and I would prefer to keep them separated. I squashed fix-only commits into the respective original commits. No backport is planned, so this shouldn't complicate anything.
* Backport on branches:
  * None, this is a quite breaking new feature.

### CODE REVIEW

File: xwiki-platform-core/pom.xml [hunk 1]
```diff
@@ -138,10 +138,23 @@
...
             -->
-            
-            
-            
-            
+
+            <revapi.differences>
+              <justification>Change in generated class of the REST model to add the
+                enforceRequiredRights property.</justification>
+              <criticality>highlight</criticality>
```
Comment (tmortagne) [inline xwiki-platform-core/pom.xml:145]: Feels like an `allowed` to me (I don't see what this change could really break in practice).

File: xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/DocumentModelBridge.java [hunk 2]
```diff
@@ -221,4 +221,16 @@ default boolean isRestricted()
     {
         return false;
     }
+
+    /**
+     * @return {@code true} if required rights defined in a {@code XWiki.RequiredRightClass} object shall be
+     * enforced, meaning that editing will be limited to users with these rights and content of this document can't
+     * use more rights than defined in the object, {@code false} otherwise
+     * @since 16.6.0RC1
```
[the remark this answers is not in the document]
Comment (michitux) [reply xwiki-platform-core/xwiki-platform-bridge/src/main/java/org/xwiki/bridge/DocumentModelBridge.java]: Changed to 16.10.0RC1.

File: xwiki-platform-core/pom.xml (hunk 1 above)
Comment (michitux) [reply xwiki-platform-core/pom.xml]: Okay, I've changed it, thank you!

**Comment by michitux** [comment]
The UI has been postponed to ensure that the API ends up in 16.10.x LTS branch, so extensions can start using the flag to enforce required rights on their extension documents. I created https://jira.xwiki.org/browse/XWIKI-22656 for the UI.

The cache that is mentioned in the PR description will be added after the merge.

**Pull request merged**